src/engine/v3/V3Interpreter.v3

```diff
@@ -1330,14 +1317,91 @@ component V3Interpreter {
 		var r = f(double.view(x));
 		val_stack.elems[val_stack.top-1] = Value.F64(u64.view(r));
 	}
+
+	// Get the next frame, or make a new one if necessary
+	private def getNextFrame() -> V3Frame {
+		if (frame == null) {
+			frame = V3Frame.new();
+			frame.prev = null;
+			frame.next = null;
+		} else if (frame.next != null) { // reuse frame object
+			frame = frame.next;
+		} else {
+			var new_frame = V3Frame.new();
+			new_frame.prev = frame;
+			frame.next = new_frame;
+			frame = new_frame;
+		}
+		num_frames++;
+		frame.index = num_frames;
+		return frame;
+	}
+	// Get the next host frame, or make a new one if necessary
+	private def getNextHostFrame() -> V3Frame {
+		frame = getNextFrame();
+		if (host_frame == null) {
+			frame.prev_host = null;
+			frame.next_host = null;
+		} else {
+			frame.prev_host = host_frame;
+			host_frame.next_host = frame;
+		}
+		host_frame = frame;
+		return host_frame;
+	}
+	private def popFrame() -> V3Frame {
+		num_frames--;
+		if (frame == host_frame) {
+			host_frame = host_frame.prev_host;
+		} else if (frame.func != null) {
+			call_stack_size--;
+		}
+		frame.hf = frame.func = frame.accessor = null;
+		frame = frame.prev;
+		return frame;
+	}
+	private def popFrames(prev_top: V3Frame) -> Array<(WasmFunction, int)> {
+		var frames: List<(WasmFunction, int)> = null;
```

Comment:
This will be more efficient if you can use a `Vector`, particularly if you can size it properly with `grow()`.

Reply:
I realise now that I store the index of the frame in the frame itself, I can just pre-size the Array directly